samza-core/src/main/java/org/apache/samza/execution/ExecutionPlanner.java

```diff
@@ -188,98 +183,82 @@ private void validateConfig() {
+
+    // Verify agreement between joined input/intermediate streams.
+    // This may involve setting partition counts of intermediate stream edges.
+    joinedStreamsGroups.forEach(ExecutionPlanner::validateJoinedStreamsGroupPartitions);
   }
 
   /**
-   * This function traverses the {@link OperatorSpec} graph to find and update mappings for all Joins reachable
-   * from this input {@link StreamEdge}.
-   * @param operatorSpec the {@link OperatorSpec} to traverse
-   * @param sourceStreamEdge source {@link StreamEdge}
-   * @param joinSpecToStreamEdges mapping from join spec to its source {@link StreamEdge}s
-   * @param streamEdgeToJoinSpecs mapping from source {@link StreamEdge} to the join specs that consumes it
-   * @param joinQ queue that contains joinSpecs where at least one of the input stream edge partitions is known.
+   * Creates a {@link JoinedStreamsGroup} whose Id is {@code groupId}, and {@link StreamEdge}s
+   * correspond to the provided {@code inputOperatorSpecs}.
    */
-  private static void findReachableJoins(OperatorSpec operatorSpec, StreamEdge sourceStreamEdge,
-      Multimap<JoinOperatorSpec, StreamEdge> joinSpecToStreamEdges,
-      Multimap<StreamEdge, JoinOperatorSpec> streamEdgeToJoinSpecs,
-      Queue<JoinOperatorSpec> joinQ, Set<JoinOperatorSpec> visited) {
-
-    if (operatorSpec instanceof JoinOperatorSpec) {
-      JoinOperatorSpec joinOperatorSpec = (JoinOperatorSpec) operatorSpec;
-      joinSpecToStreamEdges.put(joinOperatorSpec, sourceStreamEdge);
-      streamEdgeToJoinSpecs.put(sourceStreamEdge, joinOperatorSpec);
-
-      if (!visited.contains(joinOperatorSpec) && sourceStreamEdge.getPartitionCount() > 0) {
-        // put the joins with known input partitions into the queue and mark as visited
-        joinQ.add(joinOperatorSpec);
-        visited.add(joinOperatorSpec);
+  private JoinedStreamsGroup getJoinedStreamsGroup(String groupId, Iterable<InputOperatorSpec> inputOperatorSpecs,
+      JobGraph jobGraph) {
+
+    int countStreamEdgeWithSetPartitions = 0;
+    Set<StreamEdge> streamEdges = new HashSet<>();
+
+    for (InputOperatorSpec inputOpSpec : inputOperatorSpecs) {
+      StreamEdge streamEdge = jobGraph.getOrCreateStreamEdge(getStreamSpec(inputOpSpec.getStreamId(), streamConfig));
+      if (streamEdge.getPartitionCount() != StreamEdge.PARTITIONS_UNKNOWN) {
+        ++countStreamEdgeWithSetPartitions;
       }
+      streamEdges.add(streamEdge);
     }
 
-    Collection<OperatorSpec> registeredOperatorSpecs = operatorSpec.getRegisteredOperatorSpecs();
-    for (OperatorSpec registeredOpSpec : registeredOperatorSpecs) {
-      findReachableJoins(registeredOpSpec, sourceStreamEdge, joinSpecToStreamEdges, streamEdgeToJoinSpecs, joinQ,
-          visited);
+    // Determine category of stream group based on stream partition counts.
+    JoinedStreamsGroupCategory category;
+    if (countStreamEdgeWithSetPartitions == 0) {
```

Comment:
How about moving this logic within the enum and exposing it as a static helper? 
E.g. ```JoinedStreamsGroupCategory.of(int numberOfKnownPartitions);```

It will be useful for the tests as well.

Reply:
I like the idea of a static helper on the enum but let me ask a clarifying question: how is this static helper going to decide the category based on `numberOfKnownPartitions` only? Or did you mean it should receive 2 numbers, stream edge count & known partitions count?

If it's the latter then I'm not sure it's any better to write
`new JoinedStreamsGroup(groupId, streamEdges, JoinedStreamsGroupCategory.of(countStreamEdgeWithSetPartitions, streamEdges.size()))`.